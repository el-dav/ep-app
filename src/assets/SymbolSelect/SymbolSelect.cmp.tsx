import React, { FC, useState, useEffect } from 'react';
import { AutoComplete, Form } from 'antd';

enum ValidationStatus {
  ERROR = 'error',
  WARNING = 'warning',
  DEFAULT = ''
}

const companies = [
  { symbol: 'AAPL', name: 'Apple' },
  { symbol: 'MSFT', name: 'Microsoft' },
  { symbol: 'GOOGL', name: 'Google' },
  { symbol: 'VZ', name: 'Verizen' },
  { symbol: 'MMM', name: '3M' },
  { symbol: 'NFLX', name: 'Netflix' },
  { symbol: 'FB', name: 'Facebook' },
  { symbol: 'TWTR', name: 'Twitter' },
  { symbol: 'AMZN', name: 'Amazon' },
  { symbol: 'EBAY', name: 'Ebay' }
];

type Props = {
  className?: string;
  onSelect: (value: any) => void;
  value?: string;
};

const SymbolSelect: FC<Props> = ({ className, onSelect, value }) => {
  const [dataSource, setDataSource] = useState<string[]>([]);
  const [searchText, setSearchtext] = useState<string>('');
  const [helpText, setHelpText] = useState<string>('');
  const [validationStatus, setValidationStatus] = useState<ValidationStatus>(
    ValidationStatus.DEFAULT
  );

  const setValidationError = (isError: boolean) => {
    setHelpText(isError ? 'Select a valid symbol' : '');
    setValidationStatus(
      isError ? ValidationStatus.ERROR : ValidationStatus.DEFAULT
    );
  };

  const setNoSymbolsFound = () => {
    setHelpText('No symbols found for search');
    setValidationStatus(ValidationStatus.WARNING);
  };

  useEffect(() => {
    setSearchtext(value || '');
  }, [value]);

  const onSearch = (searchValue: string) => {
    setDataSource(
      searchValue
        ? companies
            .filter(
              ({ symbol, name }) =>
                symbol.toLowerCase().indexOf(searchValue.toLowerCase()) !==
                  -1 ||
                name.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1
            )
            .map(({ symbol }) => symbol)
        : []
    );
  };

  return (
    <Form.Item
      label="Symbol"
      className={`${className || ''}`}
      validateStatus={validationStatus}
      help={helpText}
    >
      <AutoComplete
        value={searchText}
        onSelect={selected => {
          onSelect(selected);
          setValidationError(false);
        }}
        placeholder="Symbol/Company"
        dataSource={dataSource}
        onSearch={onSearch}
        onChange={(text: any) => {
          if (text.length && !dataSource.length) {
            setNoSymbolsFound();
          }
          setSearchtext(text);
        }}
        onBlur={() => {
          if (companies.some(company => company.symbol === searchText)) {
            setValidationError(false);
          } else {
            setValidationError(true);
            setSearchtext('');
          }
        }}
        allowClear
      />
    </Form.Item>
  );
};

export default SymbolSelect;
