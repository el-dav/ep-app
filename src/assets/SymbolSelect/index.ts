export { default as SymbolSelect } from './SymbolSelect.cmp';
