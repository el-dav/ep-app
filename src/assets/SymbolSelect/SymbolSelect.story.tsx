import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';

import { ThemeProvider } from 'theme';

import SymbolSelect from './SymbolSelect.cmp';

const InteractiveSymbolSelect = () => {
  const [textValue, setTextValue] = useState<string>('');
  return (
    <ThemeProvider>
      <SymbolSelect
        value={textValue}
        onSelect={(value: string) => {
          setTextValue(value);
        }}
      />
    </ThemeProvider>
  );
};

storiesOf('SymbolSelect', module)
  .add('Default', () => {
    const onSelect = action('On Select');
    const value = text('Search Value', '');
    return (
      <ThemeProvider>
        <SymbolSelect value={value} onSelect={onSelect} />
      </ThemeProvider>
    );
  })
  .add('Interactive', () => <InteractiveSymbolSelect />);
