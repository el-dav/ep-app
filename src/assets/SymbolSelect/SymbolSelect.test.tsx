import React from 'react';
import renderer from 'react-test-renderer';

import { ThemeProvider } from 'theme';

import SymbolSelect from './SymbolSelect.cmp';

const onSelect = () => undefined;
const value = 'AAPL';

describe('SymbolSelect', () => {
  it('renders the Component correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider>
          <SymbolSelect onSelect={onSelect} value={value} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
