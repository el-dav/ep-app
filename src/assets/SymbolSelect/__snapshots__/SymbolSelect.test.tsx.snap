// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SymbolSelect renders the Component correctly 1`] = `
<div
  className="ant-row ant-form-item"
>
  <div
    className="ant-col ant-form-item-label"
  >
    <label
      className=""
      htmlFor={null}
      onClick={[Function]}
      title="Symbol"
    >
      Symbol
    </label>
  </div>
  <div
    className="ant-col ant-form-item-control-wrapper"
  >
    <div
      className="ant-form-item-control"
    >
      <span
        className="ant-form-item-children"
      >
        <div
          className="ant-select-show-search ant-select-auto-complete ant-select ant-select-combobox ant-select-enabled ant-select-allow-clear"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
        >
          <div
            aria-autocomplete="list"
            aria-controls="test-uuid"
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-select-selection
            ant-select-selection--single"
            role="combobox"
          >
            <div
              className="ant-select-selection__rendered"
            >
              <div
                className="ant-select-selection__placeholder"
                onClick={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "WebkitUserSelect": "none",
                    "display": "none",
                    "userSelect": "none",
                  }
                }
                unselectable="on"
              >
                Symbol/Company
              </div>
              <ul>
                <li
                  className="ant-select-search ant-select-search--inline"
                >
                  <div
                    className="ant-select-search__field__wrap"
                  >
                    <input
                      className="ant-input ant-select-search__field"
                      disabled={false}
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      type="text"
                      value="AAPL"
                    />
                    <span
                      className="ant-select-search__field__mirror"
                    >
                      AAPL
                      Â 
                    </span>
                  </div>
                </li>
              </ul>
            </div>
            <span
              className="ant-select-selection__clear"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "userSelect": "none",
                }
              }
              unselectable="on"
            >
              <i
                aria-label="icon: close-circle"
                className="anticon anticon-close-circle ant-select-clear-icon"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </i>
            </span>
            <span
              className="ant-select-arrow"
              onClick={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "userSelect": "none",
                }
              }
              unselectable="on"
            >
              <i
                aria-label="icon: down"
                className="anticon anticon-down ant-select-arrow-icon"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </i>
            </span>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
